File: udon_decompiler/analysis/transform/passes/high_level_switch_transform.py
from __future__ import annotations

from copy import deepcopy
from typing import Dict, List, Optional, cast

from udon_decompiler.analysis.ir.nodes import (
    IRBlock,
    IRBlockContainer,
    IRContainerKind,
    IRFunction,
    IRHighLevelSwitch,
    IRHighLevelSwitchSection,
    IRIf,
    IRJump,
    IRLeave,
    IRReturn,
    IRStatement,
    IRSwitch,
)
from udon_decompiler.analysis.transform.pass_base import (
    IILTransform,
    ILTransformContext,
)


class HighLevelSwitchTransform(IILTransform):
    """
    Lift low-level switch containers into high-level switch sections.

    The conversion follows ILSpy StatementBuilder's core ideas:
    - group labels by target block
    - pick a default section (largest label group when there is no explicit default)
    - inline section bodies when the target block is switch-private
    """

    def run(self, function: IRFunction, context: ILTransformContext) -> None:
        _ = context
        self._rewrite_container(function.body)

    def _rewrite_container(self, container: IRBlockContainer) -> None:
        for block in list(container.blocks):
            block.statements = self._rewrite_statement_list(
                block.statements,
                parent_container=container,
                parent_block=block,
            )

    def _rewrite_statement_list(
        self,
        statements: List[IRStatement],
        parent_container: IRBlockContainer,
        parent_block: IRBlock,
    ) -> List[IRStatement]:
        return [
            self._rewrite_statement(
                statement=statement,
                parent_container=parent_container,
                parent_block=parent_block,
            )
            for statement in statements
        ]

    def _rewrite_statement(
        self,
        statement: IRStatement,
        parent_container: IRBlockContainer,
        parent_block: IRBlock,
    ) -> IRStatement:
        if isinstance(statement, IRIf):
            statement.true_statement = self._rewrite_statement(
                statement.true_statement,
                parent_container=parent_container,
                parent_block=parent_block,
            )
            if statement.false_statement is not None:
                statement.false_statement = self._rewrite_statement(
                    statement.false_statement,
                    parent_container=parent_container,
                    parent_block=parent_block,
                )
            return statement

        if isinstance(statement, IRBlock):
            statement.statements = self._rewrite_statement_list(
                statement.statements,
                parent_container=parent_container,
                parent_block=statement,
            )
            return statement

        if isinstance(statement, IRBlockContainer):
            self._rewrite_container(statement)
            if statement.kind == IRContainerKind.SWITCH:
                return self._lift_switch_container(
                    statement,
                    parent_container,
                    parent_block,
                )
            return statement

        return statement

    def _lift_switch_container(
        self,
        switch_container: IRBlockContainer,
        parent_container: IRBlockContainer,
        parent_block: IRBlock,
    ) -> IRStatement:
        entry = switch_container.entry_block
        if entry is None or len(entry.statements) != 1:
            return switch_container

        switch_inst = entry.statements[0]
        if not isinstance(switch_inst, IRSwitch):
            return switch_container

        target_to_labels: Dict[IRBlock, List[int]] = {}
        for label_value, target in switch_inst.cases.items():
            target_to_labels.setdefault(target, []).append(label_value)
        if not target_to_labels:
            return switch_container

        explicit_default_target = switch_inst.default_target
        default_target = explicit_default_target
        largest_case_target: Optional[IRBlock] = None

        if default_target is None:
            largest_case_target = max(
                target_to_labels.items(),
                key=lambda item: len(item[1]),
            )[0]
            default_target = largest_case_target

        unique_targets: List[IRBlock] = []
        for target in target_to_labels:
            if target not in unique_targets:
                unique_targets.append(target)
        if default_target is not None and default_target not in unique_targets:
            unique_targets.append(default_target)

        incoming_counts = self._count_incoming_edges(parent_container)
        switch_ref_counts = self._count_switch_refs(switch_inst)
        common_exit = self._detect_common_exit(unique_targets)
        can_drop_common_exit_jump = parent_block in parent_container.blocks

        sections: List[IRHighLevelSwitchSection] = []
        consumed_blocks: List[IRBlock] = []

        non_default_targets = [t for t in unique_targets if t is not default_target]
        non_default_targets.sort(
            key=lambda target: min(target_to_labels.get(target, [2**31 - 1]))
        )

        for target in non_default_targets:
            labels = sorted(target_to_labels.get(target, []))
            body = self._build_section_body(
                target=target,
                switch_container=switch_container,
                parent_container=parent_container,
                incoming_counts=incoming_counts,
                switch_ref_counts=switch_ref_counts,
                common_exit=common_exit,
                can_drop_common_exit_jump=can_drop_common_exit_jump,
            )
            if (
                body.start_address == target.start_address
                and target in parent_container.blocks
            ):
                if self._is_consumable_target(
                    target=target,
                    parent_container=parent_container,
                    incoming_counts=incoming_counts,
                    switch_ref_counts=switch_ref_counts,
                    switch_container=switch_container,
                    common_exit=common_exit,
                ):
                    consumed_blocks.append(target)
            sections.append(
                IRHighLevelSwitchSection(
                    labels=labels,
                    body=body,
                    is_default=False,
                )
            )

        if default_target is not None:
            default_labels = sorted(target_to_labels.get(default_target, []))
            if (
                largest_case_target is default_target
                and explicit_default_target is None
            ):
                default_labels = []

            default_body = self._build_section_body(
                target=default_target,
                switch_container=switch_container,
                parent_container=parent_container,
                incoming_counts=incoming_counts,
                switch_ref_counts=switch_ref_counts,
                common_exit=common_exit,
                can_drop_common_exit_jump=can_drop_common_exit_jump,
            )
            if (
                default_body.start_address == default_target.start_address
                and default_target in parent_container.blocks
                and self._is_consumable_target(
                    target=default_target,
                    parent_container=parent_container,
                    incoming_counts=incoming_counts,
                    switch_ref_counts=switch_ref_counts,
                    switch_container=switch_container,
                    common_exit=common_exit,
                )
            ):
                consumed_blocks.append(default_target)

            sections.append(
                IRHighLevelSwitchSection(
                    labels=default_labels,
                    body=default_body,
                    is_default=True,
                )
            )

        for block in consumed_blocks:
            if block in parent_container.blocks:
                parent_container.blocks.remove(block)

        return IRHighLevelSwitch(
            index_expression=switch_inst.index_expression,
            sections=sections,
        )

    def _build_section_body(
        self,
        target: IRBlock,
        switch_container: IRBlockContainer,
        parent_container: IRBlockContainer,
        incoming_counts: Dict[IRBlock, int],
        switch_ref_counts: Dict[IRBlock, int],
        common_exit: Optional[IRBlock],
        can_drop_common_exit_jump: bool,
    ) -> IRBlock:
        # Most lowered switches use internal switch-container target blocks
        # ending with `leave switch`. These blocks must be inlined into
        # section bodies; emitting `jump -> target` would become dangling
        # once the switch container gets replaced by IRHighLevelSwitch.
        if target in switch_container.blocks:
            copied_statements: List[IRStatement] = list(target.statements)
            if copied_statements:
                tail = copied_statements[-1]
                if (
                    isinstance(tail, IRLeave)
                    and tail.target_container is switch_container
                ):
                    copied_statements.pop()
                elif (
                    isinstance(tail, IRJump)
                    and common_exit is not None
                    and tail.target is common_exit
                    and common_exit not in parent_container.blocks
                ):
                    copied_statements.pop()
            return IRBlock(
                statements=copied_statements,
                start_address=target.start_address,
            )

        consumable = self._is_consumable_target(
            target=target,
            parent_container=parent_container,
            incoming_counts=incoming_counts,
            switch_ref_counts=switch_ref_counts,
            switch_container=switch_container,
            common_exit=common_exit,
        )

        if not consumable:
            # Pattern seen in range-guarded lowered switches:
            # target block was inlined by an earlier pass, but we still have the
            # original target block object and can inline its payload safely.
            if (
                target not in parent_container.blocks
                and target.statements
                and isinstance(target.statements[-1], IRJump)
                and common_exit is not None
                and cast(IRJump, target.statements[-1]).target is common_exit
            ):
                copied = deepcopy(target.statements)
                # Keep the trailing jump when its destination is still emitted in
                # this container. This preserves "assign then goto exit" shape and
                # avoids duplicated fall-through assignments after the switch.
                if common_exit not in parent_container.blocks:
                    copied = copied[:-1]
                return IRBlock(
                    statements=copied,
                    start_address=target.start_address,
                )
            return IRBlock(
                statements=[IRJump(target=target)],
                start_address=target.start_address,
            )

        drop_last_jump = False
        drop_last_leave = False
        if target.statements:
            original_last = target.statements[-1]
            if (
                isinstance(original_last, IRJump)
                and can_drop_common_exit_jump
                and common_exit is not None
                and original_last.target is common_exit
            ):
                drop_last_jump = True
            elif (
                isinstance(original_last, IRLeave)
                and original_last.target_container is switch_container
            ):
                drop_last_leave = True

        # Preserve block target identity in IRJump so later label collection can
        # mark the real destination block.
        copied_statements: List[IRStatement] = list(target.statements)
        if copied_statements and (drop_last_jump or drop_last_leave):
            copied_statements.pop()

        return IRBlock(
            statements=copied_statements,
            start_address=target.start_address,
        )

    def _is_consumable_target(
        self,
        target: IRBlock,
        parent_container: IRBlockContainer,
        incoming_counts: Dict[IRBlock, int],
        switch_ref_counts: Dict[IRBlock, int],
        switch_container: IRBlockContainer,
        common_exit: Optional[IRBlock],
    ) -> bool:
        if target not in parent_container.blocks:
            return False

        if incoming_counts.get(target, 0) != switch_ref_counts.get(target, 0):
            return False

        if not target.statements:
            return True

        last = target.statements[-1]
        if isinstance(last, IRJump):
            return common_exit is not None and last.target is common_exit
        if isinstance(last, IRLeave):
            return last.target_container is switch_container
        return isinstance(last, IRReturn)

    def _detect_common_exit(self, targets: List[IRBlock]) -> Optional[IRBlock]:
        target_set = set(targets)
        counts: Dict[IRBlock, int] = {}
        for target in targets:
            if not target.statements:
                continue
            last = target.statements[-1]
            if isinstance(last, IRJump):
                exit_target = last.target
                if exit_target in target_set:
                    continue
                counts[exit_target] = counts.get(exit_target, 0) + 1

        if not counts:
            return None

        best_target, best_count = max(counts.items(), key=lambda item: item[1])
        if best_count < 2:
            return None
        return best_target

    def _count_switch_refs(self, switch_inst: IRSwitch) -> Dict[IRBlock, int]:
        counts: Dict[IRBlock, int] = {}
        for target in switch_inst.cases.values():
            counts[target] = counts.get(target, 0) + 1
        if switch_inst.default_target is not None:
            default_target = switch_inst.default_target
            counts[default_target] = counts.get(default_target, 0) + 1
        return counts

    def _count_incoming_edges(self, container: IRBlockContainer) -> Dict[IRBlock, int]:
        counts: Dict[IRBlock, int] = {}
        for block in container.blocks:
            for statement in block.statements:
                self._count_targets_in_statement(statement, counts)
        return counts

    def _count_targets_in_statement(
        self,
        statement: IRStatement,
        counts: Dict[IRBlock, int],
    ) -> None:
        if isinstance(statement, IRJump):
            counts[statement.target] = counts.get(statement.target, 0) + 1
            return

        if isinstance(statement, IRIf):
            self._count_targets_in_statement(statement.true_statement, counts)
            if statement.false_statement is not None:
                self._count_targets_in_statement(statement.false_statement, counts)
            return

        if isinstance(statement, IRSwitch):
            for target in statement.cases.values():
                counts[target] = counts.get(target, 0) + 1
            if statement.default_target is not None:
                target = statement.default_target
                counts[target] = counts.get(target, 0) + 1
            return

        if isinstance(statement, IRHighLevelSwitch):
            for section in statement.sections:
                self._count_targets_in_statement(section.body, counts)
            return

        if isinstance(statement, IRBlock):
            for nested in statement.statements:
                self._count_targets_in_statement(nested, counts)
            return

        if isinstance(statement, IRBlockContainer):
            for block in statement.blocks:
                for nested in block.statements:
                    self._count_targets_in_statement(nested, counts)

File: src/decompiler/transform/passes/high_level_switch_transform.rs
use std::collections::{HashMap, HashSet};

use crate::decompiler::Result;
use crate::decompiler::ir::{
    IrBlock, IrBlockContainer, IrContainerKind, IrFunction, IrHighLevelSwitch,
    IrHighLevelSwitchSection, IrIf, IrJump, IrLeave, IrReturn, IrStatement, IrSwitch,
};
use crate::decompiler::transform::pass_base::{ITransform, TransformContext};

pub struct HighLevelSwitchTransform;

impl ITransform for HighLevelSwitchTransform {
    fn run(
        &self,
        function: &mut IrFunction,
        _context: &mut TransformContext<'_, '_>,
    ) -> Result<()> {
        rewrite_container(&mut function.body);
        Ok(())
    }
}

fn rewrite_container(container: &mut IrBlockContainer) {
    let snapshot = container.clone();
    let mut consumed_blocks = HashSet::<u32>::new();

    for block in &mut container.blocks {
        for statement in &mut block.statements {
            let (rewritten, statement_consumed) =
                rewrite_statement(statement.clone(), &snapshot, block.start_address);
            consumed_blocks.extend(statement_consumed);
            *statement = rewritten;
        }
    }

    if !consumed_blocks.is_empty() {
        container
            .blocks
            .retain(|x| !consumed_blocks.contains(&x.start_address));
    }
}

fn rewrite_statement(
    statement: IrStatement,
    parent_container: &IrBlockContainer,
    parent_block_start: u32,
) -> (IrStatement, HashSet<u32>) {
    match statement {
        IrStatement::If(IrIf {
            condition,
            true_statement,
            false_statement,
        }) => {
            let (true_statement, mut consumed) =
                rewrite_statement(*true_statement, parent_container, parent_block_start);
            let false_statement = if let Some(x) = false_statement {
                let (rewritten, false_consumed) =
                    rewrite_statement(*x, parent_container, parent_block_start);
                consumed.extend(false_consumed);
                Some(Box::new(rewritten))
            } else {
                None
            };
            (
                IrStatement::If(IrIf {
                    condition,
                    true_statement: Box::new(true_statement),
                    false_statement,
                }),
                consumed,
            )
        }
        IrStatement::Block(mut block) => {
            let mut consumed = HashSet::<u32>::new();
            let mut rewritten = Vec::<IrStatement>::with_capacity(block.statements.len());
            for nested in block.statements {
                let (statement, nested_consumed) =
                    rewrite_statement(nested, parent_container, block.start_address);
                consumed.extend(nested_consumed);
                rewritten.push(statement);
            }
            block.statements = rewritten;
            (IrStatement::Block(block), consumed)
        }
        IrStatement::BlockContainer(mut container) => {
            rewrite_container(&mut container);
            if container.kind == IrContainerKind::Switch {
                return lift_switch_container(&container, parent_container, parent_block_start);
            }
            (IrStatement::BlockContainer(container), HashSet::new())
        }
        other => (other, HashSet::new()),
    }
}

fn lift_switch_container(
    switch_container: &IrBlockContainer,
    parent_container: &IrBlockContainer,
    parent_block_start: u32,
) -> (IrStatement, HashSet<u32>) {
    let Some(entry) = switch_container.entry_block() else {
        return (
            IrStatement::BlockContainer(switch_container.clone()),
            HashSet::new(),
        );
    };
    if entry.statements.len() != 1 {
        return (
            IrStatement::BlockContainer(switch_container.clone()),
            HashSet::new(),
        );
    }
    let IrStatement::Switch(switch_inst) = &entry.statements[0] else {
        return (
            IrStatement::BlockContainer(switch_container.clone()),
            HashSet::new(),
        );
    };

    let mut target_to_labels = HashMap::<u32, Vec<u32>>::new();
    let mut target_order = Vec::<u32>::new();
    for (label_value, target) in &switch_inst.cases {
        if !target_to_labels.contains_key(target) {
            target_order.push(*target);
        }
        target_to_labels
            .entry(*target)
            .or_default()
            .push(*label_value);
    }
    if target_to_labels.is_empty() {
        return (
            IrStatement::BlockContainer(switch_container.clone()),
            HashSet::new(),
        );
    }

    let explicit_default_target = switch_inst.default_target;
    let mut default_target = explicit_default_target;
    let mut largest_case_target = None::<u32>;

    if default_target.is_none() {
        let mut best_target = None::<u32>;
        let mut best_len = 0usize;
        for target in &target_order {
            let len = target_to_labels.get(target).map_or(0, Vec::len);
            if best_target.is_none() || len > best_len {
                best_target = Some(*target);
                best_len = len;
            }
        }
        if let Some(target) = best_target {
            largest_case_target = Some(target);
            default_target = Some(target);
        }
    }

    let mut unique_targets = target_to_labels.keys().copied().collect::<Vec<_>>();
    if let Some(default_target) = default_target
        && !unique_targets.contains(&default_target)
    {
        unique_targets.push(default_target);
    }

    let incoming_counts = count_incoming_edges(parent_container);
    let switch_ref_counts = count_switch_refs(switch_inst);
    let common_exit = detect_common_exit(parent_container, &unique_targets);
    let can_drop_common_exit_jump = parent_container
        .blocks
        .iter()
        .any(|x| x.start_address == parent_block_start);

    let mut sections = Vec::<IrHighLevelSwitchSection>::new();
    let mut consumed_blocks = HashSet::<u32>::new();

    let mut non_default_targets = unique_targets
        .iter()
        .copied()
        .filter(|x| Some(*x) != default_target)
        .collect::<Vec<_>>();
    non_default_targets.sort_by_key(|target| {
        target_to_labels
            .get(target)
            .and_then(|x| x.iter().min().copied())
            .unwrap_or(u32::MAX)
    });

    for target in non_default_targets {
        let labels = target_to_labels.get(&target).cloned().unwrap_or_default();
        let body = build_section_body(
            target,
            switch_container,
            parent_container,
            &incoming_counts,
            &switch_ref_counts,
            common_exit,
            can_drop_common_exit_jump,
        );
        if body.start_address == target
            && parent_container
                .blocks
                .iter()
                .any(|x| x.start_address == target)
            && is_consumable_target(
                target,
                parent_container,
                &incoming_counts,
                &switch_ref_counts,
                switch_container.id,
                common_exit,
            )
        {
            consumed_blocks.insert(target);
        }

        sections.push(IrHighLevelSwitchSection {
            labels,
            body,
            is_default: false,
        });
    }

    if let Some(default_target) = default_target {
        let mut default_labels = target_to_labels
            .get(&default_target)
            .cloned()
            .unwrap_or_default();
        if largest_case_target == Some(default_target) && explicit_default_target.is_none() {
            default_labels.clear();
        }

        let default_body = build_section_body(
            default_target,
            switch_container,
            parent_container,
            &incoming_counts,
            &switch_ref_counts,
            common_exit,
            can_drop_common_exit_jump,
        );
        if default_body.start_address == default_target
            && parent_container
                .blocks
                .iter()
                .any(|x| x.start_address == default_target)
            && is_consumable_target(
                default_target,
                parent_container,
                &incoming_counts,
                &switch_ref_counts,
                switch_container.id,
                common_exit,
            )
        {
            consumed_blocks.insert(default_target);
        }

        sections.push(IrHighLevelSwitchSection {
            labels: default_labels,
            body: default_body,
            is_default: true,
        });
    }

    (
        IrStatement::HighLevelSwitch(IrHighLevelSwitch {
            index_expression: switch_inst.index_expression.clone(),
            sections,
        }),
        consumed_blocks,
    )
}

fn build_section_body(
    target: u32,
    switch_container: &IrBlockContainer,
    parent_container: &IrBlockContainer,
    incoming_counts: &HashMap<u32, usize>,
    switch_ref_counts: &HashMap<u32, usize>,
    common_exit: Option<u32>,
    can_drop_common_exit_jump: bool,
) -> IrBlock {
    if let Some(target_block) = switch_container
        .blocks
        .iter()
        .find(|x| x.start_address == target)
    {
        let mut copied_statements = target_block.statements.clone();
        if let Some(tail) = copied_statements.last() {
            let drop_tail = match tail {
                IrStatement::Leave(IrLeave {
                    target_container_id,
                }) => *target_container_id == switch_container.id,
                IrStatement::Jump(IrJump { target_address }) => {
                    common_exit.is_some_and(|x| x == *target_address)
                        && !parent_container
                            .blocks
                            .iter()
                            .any(|b| b.start_address == *target_address)
                }
                _ => false,
            };
            if drop_tail {
                copied_statements.pop();
            }
        }
        return IrBlock {
            statements: copied_statements,
            start_address: target,
            should_emit_label: false,
        };
    }

    let consumable = is_consumable_target(
        target,
        parent_container,
        incoming_counts,
        switch_ref_counts,
        switch_container.id,
        common_exit,
    );

    if !consumable {
        return IrBlock {
            statements: vec![IrStatement::Jump(IrJump {
                target_address: target,
            })],
            start_address: target,
            should_emit_label: false,
        };
    }

    let Some(target_block) = parent_container
        .blocks
        .iter()
        .find(|x| x.start_address == target)
    else {
        return IrBlock {
            statements: vec![IrStatement::Jump(IrJump {
                target_address: target,
            })],
            start_address: target,
            should_emit_label: false,
        };
    };

    let mut copied_statements = target_block.statements.clone();
    if let Some(last) = copied_statements.last() {
        let drop_last_jump = matches!(
            last,
            IrStatement::Jump(IrJump { target_address })
                if can_drop_common_exit_jump && common_exit.is_some_and(|x| x == *target_address)
        );
        let drop_last_leave = matches!(
            last,
            IrStatement::Leave(IrLeave { target_container_id }) if *target_container_id == switch_container.id
        );
        if drop_last_jump || drop_last_leave {
            copied_statements.pop();
        }
    }

    IrBlock {
        statements: copied_statements,
        start_address: target,
        should_emit_label: false,
    }
}

fn is_consumable_target(
    target: u32,
    parent_container: &IrBlockContainer,
    incoming_counts: &HashMap<u32, usize>,
    switch_ref_counts: &HashMap<u32, usize>,
    switch_container_id: u32,
    common_exit: Option<u32>,
) -> bool {
    let Some(target_block) = parent_container
        .blocks
        .iter()
        .find(|x| x.start_address == target)
    else {
        return false;
    };

    if incoming_counts.get(&target).copied().unwrap_or(0)
        != switch_ref_counts.get(&target).copied().unwrap_or(0)
    {
        return false;
    }

    let Some(last) = target_block.statements.last() else {
        return true;
    };

    match last {
        IrStatement::Jump(IrJump { target_address }) => {
            common_exit.is_some_and(|x| x == *target_address)
        }
        IrStatement::Leave(IrLeave {
            target_container_id,
        }) => *target_container_id == switch_container_id,
        IrStatement::Return(IrReturn) => true,
        _ => false,
    }
}

fn detect_common_exit(parent_container: &IrBlockContainer, targets: &[u32]) -> Option<u32> {
    let target_set = targets.iter().copied().collect::<HashSet<_>>();
    let mut counts = HashMap::<u32, usize>::new();

    for target in targets {
        let Some(target_block) = parent_container
            .blocks
            .iter()
            .find(|x| x.start_address == *target)
        else {
            continue;
        };
        let Some(last) = target_block.statements.last() else {
            continue;
        };
        if let IrStatement::Jump(IrJump { target_address }) = last {
            if target_set.contains(target_address) {
                continue;
            }
            *counts.entry(*target_address).or_insert(0) += 1;
        }
    }

    let (best_target, best_count) = counts.into_iter().max_by_key(|(_, count)| *count)?;
    if best_count < 2 {
        return None;
    }
    Some(best_target)
}

fn count_switch_refs(switch_inst: &IrSwitch) -> HashMap<u32, usize> {
    let mut out = HashMap::<u32, usize>::new();
    for target in switch_inst.cases.values() {
        *out.entry(*target).or_insert(0) += 1;
    }
    if let Some(default_target) = switch_inst.default_target {
        *out.entry(default_target).or_insert(0) += 1;
    }
    out
}

fn count_incoming_edges(container: &IrBlockContainer) -> HashMap<u32, usize> {
    let mut out = HashMap::<u32, usize>::new();
    for block in &container.blocks {
        for statement in &block.statements {
            count_targets_in_statement(statement, &mut out);
        }
    }
    out
}

fn count_targets_in_statement(statement: &IrStatement, counts: &mut HashMap<u32, usize>) {
    match statement {
        IrStatement::Jump(IrJump { target_address }) => {
            *counts.entry(*target_address).or_insert(0) += 1;
        }
        IrStatement::If(IrIf {
            true_statement,
            false_statement,
            ..
        }) => {
            count_targets_in_statement(true_statement, counts);
            if let Some(false_statement) = false_statement.as_ref() {
                count_targets_in_statement(false_statement, counts);
            }
        }
        IrStatement::Switch(IrSwitch {
            cases,
            default_target,
            ..
        }) => {
            for target in cases.values() {
                *counts.entry(*target).or_insert(0) += 1;
            }
            if let Some(default_target) = default_target {
                *counts.entry(*default_target).or_insert(0) += 1;
            }
        }
        IrStatement::HighLevelSwitch(IrHighLevelSwitch { sections, .. }) => {
            for section in sections {
                for nested in &section.body.statements {
                    count_targets_in_statement(nested, counts);
                }
            }
        }
        IrStatement::Block(block) => {
            for nested in &block.statements {
                count_targets_in_statement(nested, counts);
            }
        }
        IrStatement::BlockContainer(container) => {
            for block in &container.blocks {
                for nested in &block.statements {
                    count_targets_in_statement(nested, counts);
                }
            }
        }
        _ => {}
    }
}
